gaudi: Added Gaudi Provisioning on OpenShift details (#282)
intel/intel-technology-enabling-for-openshift: gaudi/README.md

# Setting up HabanaAI Operator

## Overview
[Habana AI Operator](https://catalog.redhat.com/software/container-stacks/detail/64342b3bcbfbb9a6588ce8dd) is used to provision Intel Gaudi Accelerator with OpenShift. The steps and yaml files mentioned in this document to provision the Gaudi accelerator are based on [HabanaAI Operator for OpenShift](https://docs.habana.ai/en/latest/Orchestration/HabanaAI_Operator/index.html).

If you are familiar with the steps here to manually provision the accelerator, the Red Hat certified Operator and Ansible based [One-Click](/one_click/README.md) solution can be used as a reference to provision the accelerator automatically.

## Prerequisities

> - To Provision RHOCP cluster, Follow steps [here](/README.md).

Reviewer: Follow->follow
@chaitanya1731

Verify that the operator controller manager pod is up and running:
```
oc get pods -n habana-ai-operator

NAME                                  READY   STATUS    RESTARTS   AGE
controller-manager-6c8459d9cb-fqs8h   2/2     Running   0          25m
```

## Creating Habana AI Operator DeviceConfig Instance
To create a Habana Gaudi device plugin CR, follow the steps below.

### Create CR via web console
1.	Go to **Operator** -> **Installed Operators**.
2.	Open **HabanaAI Operator**.
3.	Navigate to tab **Device Config**.
4.	Click **Create DeviceConfig** -> set correct parameters -> Click **Create**. To set correct parameters please refer [Using RedHat OpenShift Console](https://docs.habana.ai/en/latest/Orchestration/HabanaAI_Operator/Deploying_HabanaAI_Operator.html#id2).

### Verify via web console
1.	Verify CR by checking the status of **Workloads** -> **DaemonSet** -> **habana-ai-module-device-plugin-xxxxx**.
2.	Now `DeviceConfig` is created.

### Create CR via CLI
Apply the CR yaml file:
```
oc apply -f  https://raw.githubusercontent.com/intel/intel-technology-enabling-for-openshift/main/gaudi/gaudi_device_config.yaml
```

### Verify the DeviceConfig CR is created

> Verify the Gaudi resources are up and running using below command:

Reviewer: You can use the command below to verify that the DeviceConfig CR has been created.

Apply the CR yaml file:
```
oc apply -f  https://raw.githubusercontent.com/intel/intel-technology-enabling-for-openshift/main/gaudi/gaudi_device_config.yaml
```

> ### Verify via CLI

Reviewer: ### Verify the DeviceConfig CR is created
```
oc get pod -n habana-ai-operator
NAME                                         READY   STATUS    RESTARTS        AGE
controller-manager-6586758d54-qw644          2/2     Running   0            5d5h
habana-ai-habana-runtime-bqpvp               1/1     Running   0            5d6h
habana-ai-module-device-plugin-pljkf-kxgdj   1/1     Running   0            5d6h
habana-ai-node-metrics-rghlr                 1/1     Running   0            5d6h
```

> Verify the Gaudi resources available using below command: 

Reviewer: Please move line 84 - 90 to line 94

habana-ai-node-metrics-rghlr                 1/1     Running   0            5d6h
```

## Verify Gaudi Provisioning

> After the `DeviceConfig` instance is created, it will take some time for the operator to download the Gaudi OOT driver source code and build it on-premise with the help of the KMM operator. The OOT driver module binaries will be loaded into the RHCOS kernel on each node with Gaudi cards labelled by NFD. Then, the Gaudi device plugin can advertise the Gaudi resources listed in the table for the pods on OpenShit to use. Run the command below to check the availability of Gaudi resources:

Reviewer: instance -> CR

```

### Verify the DeviceConfig CR is created
Verify the Gaudi resources are up and running using below command:
```
oc get pod -n habana-ai-operator

NAME                                         READY   STATUS    RESTARTS        AGE
controller-manager-6586758d54-qw644          2/2     Running   0            5d5h
habana-ai-habana-runtime-bqpvp               1/1     Running   0            5d6h
habana-ai-module-device-plugin-pljkf-kxgdj   1/1     Running   0            5d6h
habana-ai-node-metrics-rghlr                 1/1     Running   0            5d6h
```

## Verify Gaudi Provisioning
After the `DeviceConfig` instance is created, it will take some time for the operator to download the Gaudi OOT driver source code and build it on-premise with the help of the KMM operator. The OOT driver module binaries will be loaded into the RHCOS kernel on each node with Gaudi cards labelled by NFD. Then, the Gaudi device plugin can advertise the Gaudi resources listed in the table for the pods on OpenShit to use. Run the command below to check the availability of Gaudi resources:
```
oc describe node/<NODE> | grep habana.ai/gaudi

> 

Reviewer: oc describe node  | grep habana.ai/gaudi

  habana.ai/gaudi:    8
  habana.ai/gaudi:    8
```

> Alternatively, you can also check the status of the `DeviceConfig` CR like below: 

Reviewer: Can we move line 96 to 109 to line 87 the "Verify Gaudi Provisioning" section

# Setting up HabanaAI Operator

> ## Overview

Author: Made the requested changes.

The steps mentioned in this document to deploy and configure [Habana AI Operator](https://catalog.redhat.com/software/container-stacks/detail/64342b3bcbfbb9a6588ce8dd) on Red Hat OpenShift are consolidated and referenced from [HabanaAI Operator for OpenShift](https://docs.habana.ai/en/latest/Orchestration/HabanaAI_Operator/index.html).

## Prerequisities
- Provisioned RHOCP cluster. Follow steps [here](/README.md).

> 

Author: Made the requested changes.

> ## Setting up OpenShift Environment

Author: Made the requested changes.

- Setup Machine Configuration to update Firmware search path using below command. Please note that this step will reboot the nodes.

> ```

Reviewer: Can we change it
## Update Kernel firmware search path with MCO
Notes: This step will reboot the nodes, it is recommended to do this in the first step.  

The default Kernel firmware search path /lib/firmware in RHCOS is not writable. Command below can be used to add path /var/lib/fimware into the firmware search path list.

Author: Made the requested changes.

oc apply -f https://raw.githubusercontent.com/intel/intel-technology-enabling-for-openshift/main/gaudi/gaudi_firmware_path.yaml
```
 
- Setup Node Feature Discovery (NFD) instance using below command:

> ```

Reviewer: Can we change it to
## Label Gaudi accelerator nodes With NFD
NFD operator can be used to configure NFD to automatically detect the Gaudi accelerators and label the nodes for the flowing provisioning steps.

Author: Made the requested changes.

oc apply -f https://raw.githubusercontent.com/intel/intel-technology-enabling-for-openshift/main/gaudi/gaudi_nfd_instance_openshift.yaml
```
Verify NFD has labelled the node correctly:
```
oc get no -o json | jq '.items[].metadata.labels' | grep pci-1da3

"feature.node.kubernetes.io/pci-1da3.present": "true",
```

> NFD discovers underlying Habana Gaudi using its PCI device class and the vendor. Here `"feature.node.kubernetes.io/pci-1da3.present": "true"`, represents node has PCI card from the vendor 1da3 which is HabanaLabs. 

Reviewer: Do you think detects is better than discovers? "
discovers -> detects

Habana Gaudi -> Gaudi Accelerator

Could we remove
"Here `"feature.node.kubernetes.io/pci-1da3.present": "true"`, represents node has PCI card from the vendor 1da3 which is HabanaLabs. "

Author: Made the requested changes.

## Install HabanaAI Operator on Red Hat OpenShift
### Installation via web console

> Follow the steps below to install Intel Device Plugins Operator using OpenShift web console:

Reviewer: Intel Device Plagins Operator -> Habana AI Operator

Author: Made the requested changes.

1.	In the OpenShift web console, navigate to **Operator** -> **OperatorHub**.
2.	Search for **HabanaAI Operator** in all items field -> Click **Install**.
### Verify Installation via web console
1.	Go to **Operator** -> **Installed Operators**.
2.	Verify that the status of the operator is **Succeeded**.

### Installation via command line interface (CLI)

> Apply the [install_operator.yaml](/device_plugins/install_operator.yaml) file:

Reviewer: Can we remove this line?

Author: Made the requested changes.

```
oc apply -f https://raw.githubusercontent.com/intel/intel-technology-enabling-for-openshift/main/gaudi/gaudi_install_operator.yaml
```

### Verify Installation via CLI
Verify that the operator controller manager pod is up and running:
```
oc get pod -n habana-ai-operator

NAME                                  READY   STATUS    RESTARTS   AGE
controller-manager-6c8459d9cb-fqs8h   2/2     Running   0          25m
```

> ## Resources Provided by Habana Gaudi Device Plugin

Reviewer: Before this paragraph, should we add some more detail for user to check whether the provisioning is proper? Since the step above only shows that the operator is up and running, we do not know whether the OOT driver and device plugin has been loaded. OR we can just hide these detail and have a simple way to check the ready of provisioning? like check the number of   habana.ai/gaudi resources? 

And this paragraph should be move the end of the doc.

Author: This is just the installation of the operator. OOT drivers and gaudi device plugin is not yet created. 
It is in the following section - Creating Habana AI Operator DeviceConfig Instance.

Reviewer: Please move the resource table to the end of the doc

> The resources are the user interface for customers to claim and consume the hardware features provided by Gaudi Device Plugins from the user pods. See below table for the details:

Reviewer: Can we change to 
Resources Provided

We shouldn't expose the extra information to the user.

Author: Can you check the modified information.

> The resources are the user interface for customers to claim and consume the hardware features provided by Gaudi Device Plugins from the user pods. See below table for the details:

Reviewer: Let's remove "Gaudi Device Plugin"

Author: Made the requested changes.

| Feature | Resources | Description |
| ------- | --------- | ----------- |

> | Habana Gaudi | `habana.ai/gaudi` | Habana Gaudi Card for user pod to claim | 

Reviewer: The Number of Habana Gaudi Card resources ready for claiming

Author: Made the requested changes.

> ## Creating Habana Gaudi Device Plugin custom resource (CR)

Reviewer: ## Creating Habana AI Operator DeviceConfig Instance

Author: Made the requested changes.

> To create an Intel SGX device plugin CR, follow the steps below.

Reviewer 2: ```suggestion
To create a Habana Gaudi device plugin CR, follow the steps below.
```

Author: good catch.. updated. @hershpa

### Create CR via web console
1.	Go to **Operator** -> **Installed Operators**.
2.	Open **HabanaAI Operator**.
3.	Navigate to tab **Device Config**.
4.	Click **Create DeviceConfig** -> set correct parameters -> Click **Create**. To set correct parameters please refer [Using RedHat OpenShift Console](https://docs.habana.ai/en/latest/Orchestration/HabanaAI_Operator/Deploying_HabanaAI_Operator.html#id2).

> 5.	Optional: If you want to  make any customizations, select YAML view and edit the details. Once you are done, click **Create**.

Reviewer: We can remove this line.

Author: Made the requested changes.

> # Setting up HabanaAI Operator

Author: Already added the link to the OOT driver information from habana documentation on line 65.

```
NFD discovers underlying Habana Gaudi using its PCI device class and the vendor. Here `"feature.node.kubernetes.io/pci-1da3.present": "true"`, represents node has PCI card from the vendor 1da3 which is HabanaLabs. 

## Install HabanaAI Operator on Red Hat OpenShift
### Installation via web console
Follow the steps below to install Intel Device Plugins Operator using OpenShift web console:
1.	In the OpenShift web console, navigate to **Operator** -> **OperatorHub**.
2.	Search for **HabanaAI Operator** in all items field -> Click **Install**.
### Verify Installation via web console
1.	Go to **Operator** -> **Installed Operators**.
2.	Verify that the status of the operator is **Succeeded**.

### Installation via command line interface (CLI)
Apply the [install_operator.yaml](/device_plugins/install_operator.yaml) file:
```
oc apply -f https://raw.githubusercontent.com/intel/intel-technology-enabling-for-openshift/main/gaudi/gaudi_install_operator.yaml
```

### Verify Installation via CLI
Verify that the operator controller manager pod is up and running:
```
oc get pod -n habana-ai-operator

NAME                                  READY   STATUS    RESTARTS   AGE

> controller-manager-6c8459d9cb-fqs8h   2/2     Running   0          25m

Author: This is just to verify the operator installation and verify that the controller manager pod is up and running.. Gaudi resources are created once the device plugin is created, this is mentioned in the below section to verify the gaudi resources..

# Setting up HabanaAI Operator

## Overview
[Habana AI Operator](https://catalog.redhat.com/software/container-stacks/detail/64342b3bcbfbb9a6588ce8dd) is used to provision Intel Gaudi Accelerator with OpenShift. The steps and yaml files mentioned in this document to provision the Gaudi accelerator are based on [HabanaAI Operator for OpenShift](https://docs.habana.ai/en/latest/Orchestration/HabanaAI_Operator/index.html).

If you are familiar with the steps here to manually provision the accelerator, the Red Hat certified Operator and Ansible based [One-Click](/one_click/README.md) solution can be used as a reference to provision the accelerator automatically.

## Prerequisities
- Provisioned RHOCP cluster. Follow steps [here](/README.md).
- Installed NFD Operator, follow steps [here](/nfd/README.md).

> - Installed KMM Operator, follow steps [here](/kmmo/README.md).

Reviewer: Please use the same format.

| Feature | Resources | Description |
| ------- | --------- | ----------- |
| Habana Gaudi | `habana.ai/gaudi` | Number of Habana Gaudi Card resources ready to claim | 

## Creating Habana AI Operator DeviceConfig Instance
To create a Habana Gaudi device plugin CR, follow the steps below.

### Create CR via web console
1.	Go to **Operator** -> **Installed Operators**.
2.	Open **HabanaAI Operator**.
3.	Navigate to tab **Device Config**.
4.	Click **Create DeviceConfig** -> set correct parameters -> Click **Create**. To set correct parameters please refer [Using RedHat OpenShift Console](https://docs.habana.ai/en/latest/Orchestration/HabanaAI_Operator/Deploying_HabanaAI_Operator.html#id2).

### Verify via web console
1.	Verify CR by checking the status of **Workloads** -> **DaemonSet** -> **habana-ai-module-device-plugin-xxxxx**.
2.	Now `DeviceConfig` is created.

### Create CR via CLI
Apply the CR yaml file:
```
oc apply -f  https://raw.githubusercontent.com/intel/intel-technology-enabling-for-openshift/main/gaudi/gaudi_device_config.yaml
```

### Verify via CLI
Verify that the device plugin CR is ready:

> ```

Reviewer: device plugin -> DeviceConfig

Reviewer: How about the comments here? @chaitanya1731

oc get DeviceConfig -n habana-ai-operator --show-kind=true
NAME                               AGE
deviceconfig.habana.ai/habana-ai   2d8h

> ```

Reviewer: Can we add one more section to verify that the Provisioning is successful?

## Verifying Gaudi provisioning
After the DeviceConfig instance is created, It will take some time for the operator to download the Gaudi OOT driver source code and build it on-premise with the help of the KMM operator. The OOT driver module binaries will be loaded into the RHCOS kernel on each node with Gaudio Cards labelled by NFD. Then, the Gaudi device plugin can advertise the Gaudi resources listed in the table for the pods on OpenShit to use.
Use the command below to check the availability of Gaudi resources.
@chaitanya1731  please add the command to check the Gaudi resources, and move the table and related paragram here. 

@chaitanya1731 It is great if we can add command to check  nodeMetrics and runtime. 
Thanks,
